src/v/raft/consensus.cc

```diff
@@ -420,6 +421,97 @@ void consensus::dispatch_recovery(follower_index_metadata& idx) {
     });
 }
 
+ss::future<result<model::offset>> consensus::linearizable_barrier() {
+    using ret_t = result<model::offset>;
+    struct state_snapshot {
+        model::offset linearizable_offset;
+        model::term_id term;
+    };
+
+    std::optional<ss::semaphore_units<>> u = co_await _op_lock.get_units();
+
+    if (_vstate != vote_state::leader) {
+        co_return result<model::offset>(make_error_code(errc::not_leader));
+    }
+    // store current commit index
+    auto cfg = config();
+    auto dirty_offset = _log.offsets().dirty_offset;
+    /**
+     * Dispatch round of heartbeats
+     */
+
+    absl::flat_hash_map<vnode, follower_req_seq> sequences;
+    std::vector<ss::future<>> send_futures;
+    send_futures.reserve(cfg.unique_voter_count());
+    cfg.for_each_voter([this, dirty_offset, &sequences, &send_futures](
+                         vnode target) {
+        // do not send request to self
+        if (target == _self) {
+            return;
+        }
+        // prepare empty request
+        append_entries_request req(
+          _self,
+          target,
+          meta(),
+          model::make_memory_record_batch_reader(
+            ss::circular_buffer<model::record_batch>{}),
+          append_entries_request::flush_after_append::no);
+        auto seq = next_follower_sequence(target);
+        sequences.emplace(target, seq);
+
+        update_node_append_timestamp(target);
+        vlog(
+          _ctxlog.trace, "Sending empty append entries request to {}", target);
+        auto f
+          = _client_protocol
+              .append_entries(
+                target.id(),
+                std::move(req),
+                rpc::client_opts(_replicate_append_timeout + clock_type::now()))
+              .then([this, id = target.id(), seq, dirty_offset](
+                      result<append_entries_reply> reply) {
+                  process_append_entries_reply(id, reply, seq, dirty_offset);
+              });
+
+        send_futures.push_back(std::move(f));
+    });
+    // release semaphore
+    // snapshot taken under the semaphore
+    state_snapshot snapshot{
+      .linearizable_offset = _commit_index, .term = _term};
+    u.reset();
+
+    // wait for responsens in background
+    (void)ss::with_gate(_bg, [futures = std::move(send_futures)]() mutable {
+        return ss::when_all_succeed(futures.begin(), futures.end());
+    });
+
+    auto majority_sequences_updated = [&cfg, &sequences, this] {
+        return cfg.majority([this, &sequences](vnode id) {
+            if (id == _self) {
+                return true;
+            }
+            if (auto it = _fstats.find(id); it != _fstats.end()) {
+                return it->second.last_received_seq >= sequences[id];
+            }
+            return false;
+        });
+    };
+
+    // we do not hold the lock while waiting
+    co_await _follower_reply.wait(
```

Comment:
what action prior to this `wait` arranges for an eventual `signal/broadcast`? since there are some scheduling events prior to this `wait` this looks like a race condition. remember that that condition variables drop signals/broadcasts if no waiter is present at the time the signal arrives.

Reply:
Michal pointed out that when a signal is dropped that the next heartbeat will resignal the condition variable so worst case scenario is a heartbeat extra delay.